File: lc_py/lcPy/stack/nextGreaterElements.py
from typing import List


class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        ans = [-1] * len(nums)
        n = len(nums)
        st = []

        for i in range(2 * n - 1, -1 , -1):
            curr = nums[i % n]
            while st and nums[st[-1]] <= curr:
                st.pop()
            if st and i < n: ans[i] = nums[st[-1]]
            st.append(i % n)
        return ans

File: lc_cpp/lcCpp/stack/nextGreaterElements.cpp
#include <vector>
#include <stack>
using namespace std;
class Solution {
    public:
        vector<int> nextGreaterElements(vector<int>& nums) {
            int n = nums.size();
            vector<int> ans(n, -1);
            stack<int> st;
    
            for (int i = 2 * n - 1; i >= 0; --i) {
                int curr = nums[i % n];
                while (!st.empty() && nums[st.top()] <= curr) {
                    st.pop();
                }
                if (i < n && !st.empty()) ans[i] = nums[st.top()];
    
                st.push(i % n);
            }
            return ans;
        }
    };